### PR TITLE
generate tokens_to_generate for all items in a batch

### PR DESCRIPTION
**Pull request by arendu**
# What does this PR do ?

ensure that all examples in a batch generate only `tokens_to_generate` number of tokens.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/text_generation_utils.py [hunk 1]
```diff
@@ -748,6 +750,9 @@ def sample_sequence_batch(
 
                 # Clamp the predicted out of vocabulary tokens
                 prev = torch.clamp(prev, max=tokenizer.vocab_size - 1)
+                # force EOS token when tokens_to_generate has been reached for an item in the batch.
+                prev[context_length >= maxlens] = tokenizer.eos_id
```
Comment (odelalleau) [inline nemo/collections/nlp/modules/common/text_generation_utils.py:754]: This looks ok to me, but there might be an edge case if we happen to use a text generation strategy that doesn't consider `EOS` as the end of generation (the computed length would be incorrect). Is this something that needs to be supported? If yes, I think two things could be done:

1. Do something like `done_token = done_token | (context_length == maxlens).byte()` after the definition of `done_token` below (to be sure we consider generation to be done even if the generation strategy ignores the `EOS` token)
2. Use `prev[context_length == maxlens] = tokenizer.eos_id` here (no need for `>=` since `is_done` will be set for future steps, which will force `EOS` on l.759)

File: nemo/collections/nlp/modules/common/text_generation_utils.py [hunk 2]
```diff
@@ -697,9 +697,11 @@ def sample_sequence_batch(
         all_generated_indices = None  # used to track all generated indices
         # Generate enough tokens for the longest sequence
         maxlen = tokens_to_generate + context_lengths.max().item()
-
         maxlen = inference_strategy.clip_max_len(maxlen)
 
+        maxlens = tokens_to_generate + context_lengths  # used to force EOS for shorter contexts in a batch.
```
Comment (odelalleau) [inline nemo/collections/nlp/modules/common/text_generation_utils.py:702]: Do we need to also implement a similar logic in `tab_sample_sequence_batch()`? (by the way this function seems to already have drifted away a bit from `sample_sequence_batch()`, and I'm not sure this is intended)

**Review by odelalleau — COMMENTED** [review]
Looks good to me but a couple of things that might require a closer look

File: nemo/collections/nlp/modules/common/text_generation_utils.py (hunk 1 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/text_generation_utils.py:754]: what about using the padding token?

File: nemo/collections/nlp/modules/common/text_generation_utils.py (hunk 2 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/text_generation_utils.py:702]: yes. let 's fix it in the `tab_sample_sequence_batch` which is used to tabular data generation.

Comment (arendu) [reply nemo/collections/nlp/modules/common/text_generation_utils.py:702]: updated for tab_sample

File: nemo/collections/nlp/modules/common/text_generation_utils.py (hunk 1 above)
Comment (odelalleau) [reply nemo/collections/nlp/modules/common/text_generation_utils.py:754]: > Is this something that needs to be supported?

As discussed, I'm ok not supporting it for now.

> what about using the padding token?

As discussed, there are some additional subtleties in properly handling padded inputs => personally I'm ok with not doing it in this PR.

**Comment by arendu** [comment]
jenkins

**Comment by ericharper** [comment]
jenkins

**Comment by arendu** [comment]
jenkins
